### ISSUE TITLE
[BUG] - SmartCars3 crashes every time 

### ISSUE THREAD

**Issue by atanneberg** [invernyx/smartcars-3-bugs#374]
### Describe the bug

After a fresh start of XP12 and a fresh start of SC3, connected to my VA and to XP12, flight started SC3 uses 162 MB of RAM  which seems normal.
 
But to me it looks like there is a memory leak in the SC3 App as it constantly increases the RAM-usage over the runtime and it finally crashed during flight at around 14Gb RAM (!!!) usage which is more than the XP12 itself with orthos and other add-ons needs ( XP12 was using about 9-10GB during this flight)
 
My Mac has 32GB of RAM and the crash happened shortly after the physical RAM was completely used and the system started to use memory swapping.
 
Attached you will find the MacOS Crash-Log for SC3 – the SC3 I am using is the latest available from your website, version 3.1.1 which is still Intel only , so it uses rosetta to run.
 
I really would like to use SC3, but as it crashes every flight it is unusable at the moment.

### How do you reproduce this bug?

It is happening every time. After SC3 has occupied all available memory during flight it crashes 

### Expected behavior

No excessive memory use during run-time, not crashing 

### Operating system

MacOS Sonoma

### Community airline

Walker Air

### smartCARS Version

3.1.1

### Plugins installed

chat, flight center, flight tracking, logbook, map, simbrief

### Additional context

There was already a mail-thread to the ticket no: Ticket 18705 with further information, no solution provided so far but ticket closed

**Comment by atanneberg** [comment]
[SC3 MacOS Crash-Log.txt](https://github.com/invernyx/smartcars-3-bugs/files/14088419/SC3.MacOS.Crash-Log.txt)

**Comment by GenericNerd** [comment]
@Thien42 can you look into this when you get the time please?

**Comment by tinroof** [comment]
This appears to be related to issue #352. Can SC3 be launched from a command line in Mac? Doing so in Linux is informative prior to the SC3 crash.

**Comment by atanneberg** [comment]
> This appears to be related to issue #352. Can SC3 be launched from a command line in Mac? Doing so in Linux is informative prior to the SC3 crash.

I tried to replicate the crash today while having started SC3 from the terminal. As always as soon as you start the flight the RAM use of SC3 increases permanently and the crash happened as expected after the RAM use exceeded more than 12 GB in activity monitor - a lot more than the whole XP12 was using at that time...
Additionally the CPU usage of SC3 is above 50% during the whole flight which also seems to be way to much for it.

Attached two logs; one is the MacOS Crash Report for SC3, the other is the terminal ouput of the whole flight until the crash happened.

[Error Report SC3_Feb_05_2024.txt](https://github.com/invernyx/smartcars-3-bugs/files/14167884/Error.Report.SC3_Feb_05_2024.txt)

[Terminal_Output SC3_run_via-Terminal_Feb_05_2024.txt](https://github.com/invernyx/smartcars-3-bugs/files/14167888/Terminal_Output.SC3_run_via-Terminal_Feb_05_2024.txt)

**Comment by GenericNerd** [comment]
Per @Thien42's recent testing, we can confirm that there is a memory consumption issue which appears to only be present **in MacOS**. This is important since the same increase in memory consumption is not present in Windows or Linux under similar steps to reproduce.

With this being said, we believe there is an issue with a memory leak in one of our dependencies specifically on MacOS (the most likely being Electron). We'll keep this thread up to date as there are developments.

**Comment by atanneberg** [comment]
Thanks for the confirmation

**Comment by GenericNerd** [comment]
@atanneberg It's absolutely fine, I'm sorry we really don't have anything else to say at the moment!

**Comment by tinroof** [comment]
I've reported this elsewhere however for completeness I also add this here.

I have recently been able to isolate the SC3 CTD behavior to tracking flights whilst logged into my TFDI Pro Account.  With the release of 3.2 the time frame for CTD has yet to exceed an hour. Shortest being 20 mins. 

If I skip logging into my TFDI account and instead log in to a community account, flights can be tracked indefinitely as far as I can tell. I just completed a 10+ hour long haul without incident flying with Compass Virtual. Until I established this I was finding myself unable to fly with SC3 on my own and for a VA. At least with this knowledge I can return to contributing to a VA.

I am running Linux and others reporting this are running MacOS. I have been informed by a fellow Windows pilot that this CTD issue had been plaguing his attempts to record flights. He also benefited by not logging into his TFDI account. So OS may not be a contributing factor here.

Hopefully this proves informative to the developers 

Rusty

**Comment by atanneberg** [comment]
I can confirm the observation of @tinroof on a Mac now - as long as I skip the TFDI-login at the start of SC3 all is ok and SC3 tracks for hours without excessive memory consumption.  Did several flights in the meantime with SC3 for a virtual airline without problem.

**Comment by Thien42** [comment]
@atanneberg @tinroof very sorry for the delay and thank you so much for the investigation work.

We have now released an Apple silicon version of the app at https://smartcars.tfdidesign.com/central/download
We believe this should solve the memory issue you have been having.

Please remove the current app and replace it with the new one.

Let us know if you are still having app crashes.

**Comment by Casioalpha** [comment]
Hello, I was directed here by my VA to resolve my MAC arm issue. Every time I try to download it advises it is damaged and cannot be open - please move to bin. (have tried over the past 24 hours). Are you able to assist?

**Comment by tinroof** [comment]
Might be presumptuous to have marked issue #374 closed. At least it isn't as far as running SmartCARS 3 on a linux system is concerned.

As of 18 July 2024 I removed my version of smartCARS 3. I have since downloaded and installed the most recent release "smartCARS 3-linux-x64-3.2.2".

I then run smartCARS 3 executable binary in a command window so I can monitor STDOUT and ERROUT. In an attempt to reproduce the error reported in this thread, I logged in to my TFDi account, then my community VA. Setup an A320 flight from WSSS to VTSP and started the flight in X-Plane 12.1.1.  and commenced SC3 tracking.

In less than 60 minutes the following message was output to command console.
```
`info: Connected to pro-monitoring server {"timestamp":"2024-07-18T07:42:20.573Z"}
info: Setting current plugin {"handler":"set-current-plugin","plugin":"com.tfdidesign.flight-center","source":"realTimeAPI","timestamp":"2024-07-18T07:42:24.224Z"}
info: Received notification {"handler":"notification","notification":{"id":"1f6d2e0e-26b0-4f10-9d06-bbd0dd00e668","message":"Flight created","quiet":true,"source":"com.tfdidesign.flight-center","timestamp":1721288567811,"toast":true,"type":"success"},"source":"realTimeAPI","timestamp":"2024-07-18T07:42:47.811Z"}
info: Blocking close for flight-in-progress {"timestamp":"2024-07-18T07:43:04.669Z"}
info: Flight started {"timestamp":"2024-07-18T07:43:04.689Z"}
info: Received notification {"handler":"notification","notification":{"id":"9f465208-79b9-4417-bd1c-29a492a8e923","message":"Flight started successfully.","quiet":true,"source":"com.tfdidesign.flight-tracking","timestamp":1721288584689,"toast":true,"type":"success"},"source":"realTimeAPI","timestamp":"2024-07-18T07:43:04.689Z"}

<--- Last few GCs --->

[36583:0x2a9c002ec000]  1561686 ms: Mark-Compact (reduce) 222.2 (279.6) -> 222.2 (254.4) MB, 326.36 / 0.00 ms  (average mu = 0.770, current mu = 0.000) external memory pressure; GC in old space requested
[36583:0x2a9c002ec000]  1562053 ms: Mark-Compact (reduce) 222.2 (254.4) -> 222.2 (254.4) MB, 367.10 / 0.00 ms  (average mu = 0.599, current mu = 0.000) external memory pressure; GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: v8::ArrayBuffer::NewBackingStore Allocation failed - process out of memory
zsh: IOT instruction (core dumped)  /opt/smartCARS/smartCARS3
```
I intend to also undertake a flight tracked with SC3 logging in to just community and NOT first logging into my Pro TFDi account first. I am expecting this to function for unlimited periods of time without the above error being encountered.

Rusty

**Comment by Thien42** [comment]
@Casioalpha I could confirm the issue with an ARM Mac, we will try to get this sorted

**Comment by Thien42** [comment]
@tinroof see my comment on https://github.com/invernyx/smartcars-3-bugs/issues/386

If you have additional findings, maybe put them in #386

**Comment by tinroof** [comment]
As I mentioned in my previous comment, I completed another tracked flight. Again I launched SC3 client from a command prompt. On this occasion I tracked a community only flight. That is, I skipped past the initial invitation to log in to my TFDi PRO account and instead selected my community VA. For this flight I flew the same A320NEO from VTSP to VYYY. This flight took 75 minutes and was tracked and PIREP logged without any issue.

There appears to be some memory leak when the "tracking plugin" attempts to retain PRO flight data for later storage after the completion of a flight. Unfortunately the tracking duration is so short that I have yet to track a short enough duration flight for the client to complete and upload.

My next VA flight is scheduled as a 13+ hour long haul, so I will do the above again and confirm a community without PRO flight can be tracked without concern. I then intend to fly a contrived short duration hop to confirm the client can indeed complete short hops in PRO & community mode.

I noted in the earlier command prompt PRO flight trace that TFDi SC3 is monitoring PRO flights. I would be curious to know what is being collected, if anything, given the crash is being brought about by a memory exception.

Rusty
